### PR TITLE
Add `Sequence` support for `shape` arguments

### LINKED ISSUE #7427: type hint of `shape` of `initializers.generate_array` is too broad
`chainerx.empty` does not support general sequences of ints.

https://github.com/chainer/chainer/blob/af7ea7895e475ea2bfba5d3c1b3065ca16d8e92c/chainer/types.py#L24-L27

https://github.com/chainer/chainer/blob/af7ea7895e475ea2bfba5d3c1b3065ca16d8e92c/chainer/initializers/__init__.py#L28-L29

### Code to reproduce
```python
import chainer


v = chainer.Parameter(chainer.initializers.Normal(1.))
v.to_device('native:0')
v.initialize(shape=[3, 4])
```

### Error messages, stack traces, or logs
```
Traceback (most recent call last):
  File "new_issue.py", line 6, in <module>
    v.initialize(shape=[3, 4])
  File "/Users/tos/GitHub/chainer/chainer/variable.py", line 1878, in initialize
    self.initializer, shape, xp, device=device)
  File "/Users/tos/GitHub/chainer/chainer/initializers/__init__.py", line 73, in generate_array
    array = chainerx.empty(shape, dtype=dtype, device=chx_device)
TypeError: empty(): incompatible function arguments. The following argument types are supported:
    1. (shape: tuple, dtype: handle=None, device: handle=None) -> chainerx.ndarray
    2. (shape: int_, dtype: handle=None, device: handle=None) -> chainerx.ndarray

Invoked with: [3, 4]; kwargs: dtype=dtype('float32'), device=native:0
```

### PR DESCRIPTION
**Pull request by okapies**
- Add `Sequence[int]` support for `shape` arguments in ChainerX functions, which only accept `int` and `Tuple[int]` at the moment.
- Fix tests to check if `shape` is a list of integers

Fixes #7427.

## Note
I added `#include <pybind11/stl.h>` to `chainerx_cc/chainerx/python/shape.h` to avoid the following compilation error. It maybe related to pybind/pybind11#1055.

<details><summary>error message</summary>
<p>

```
    ...
    [ 98%] Building CXX object chainerx/python/CMakeFiles/_core.so.dir/shape.o
    [100%] Linking CXX shared module /home/user/chainer/chainerx/_core.so
    /home/user/chainer/build/temp.linux-x86_64-3.6/pybind11/include/pybind11/stl.h:179:49: error: type ‘struct type_caster’ violates one definition rule [-Werror=odr]
     template <typename Type, typename Alloc> struct type_caster<std::vector<Type, Alloc>>
                                                     ^
    /home/user/chainer/build/temp.linux-x86_64-3.6/pybind11/include/pybind11/cast.h:866:56: note: a type with different bases is defined in another translation unit
     template <typename type, typename SFINAE = void> class type_caster : public type_caster_base<type> { };
                                                            ^
    lto1: all warnings being treated as errors
    lto-wrapper: fatal error: /usr/bin/c++ returned 1 exit status
    compilation terminated.
    /usr/bin/ld: error: lto-wrapper failed
    collect2: error: ld returned 1 exit status
    chainerx/python/CMakeFiles/_core.so.dir/build.make:387: recipe for target '/home/user/chainer/chainerx/_core.so' failed
    make[3]: *** [/home/user/chainer/chainerx/_core.so] Error 1
    CMakeFiles/Makefile2:614: recipe for target 'chainerx/python/CMakeFiles/_core.so.dir/all' failed
    make[2]: *** [chainerx/python/CMakeFiles/_core.so.dir/all] Error 2
    CMakeFiles/Makefile2:626: recipe for target 'chainerx/python/CMakeFiles/_core.so.dir/rule' failed
    make[1]: *** [chainerx/python/CMakeFiles/_core.so.dir/rule] Error 2
    Makefile:281: recipe for target '_core.so' failed
    make: *** [_core.so] Error 2
    /home/user/.pyenv/versions/3.6.5/lib/python3.6/distutils/dist.py:261: UserWarning: Unknown distribution option: 'long_description_content_type'
      warnings.warn(msg)
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/home/user/chainer/setup.py", line 197, in <module>
        setup(**setup_kwargs)
      File "/home/user/.pyenv/versions/3.6.5/lib/python3.6/site-packages/setuptools/__init__.py", line 129, in setup
        return distutils.core.setup(**attrs)
      File "/home/user/.pyenv/versions/3.6.5/lib/python3.6/distutils/core.py", line 148, in setup
        dist.run_commands()
      File "/home/user/.pyenv/versions/3.6.5/lib/python3.6/distutils/dist.py", line 955, in run_commands
        self.run_command(cmd)
      File "/home/user/.pyenv/versions/3.6.5/lib/python3.6/distutils/dist.py", line 974, in run_command
        cmd_obj.run()
      File "/home/user/.pyenv/versions/3.6.5/lib/python3.6/site-packages/setuptools/command/develop.py", line 36, in run
        self.install_for_development()
      File "/home/user/.pyenv/versions/3.6.5/lib/python3.6/site-packages/setuptools/command/develop.py", line 136, in install_for_development
        self.run_command('build_ext')
      File "/home/user/.pyenv/versions/3.6.5/lib/python3.6/distutils/cmd.py", line 313, in run_command
        self.distribution.run_command(command)
      File "/home/user/.pyenv/versions/3.6.5/lib/python3.6/distutils/dist.py", line 974, in run_command
        cmd_obj.run()
      File "/home/user/chainer/chainerx_build_helper.py", line 37, in run
        self.build_extension(ext)
      File "/home/user/chainer/chainerx_build_helper.py", line 88, in build_extension
        ['cmake', '--build', '.'] + build_args, cwd=self.build_temp)
      File "/home/user/.pyenv/versions/3.6.5/lib/python3.6/subprocess.py", line 291, in check_call
        raise CalledProcessError(retcode, cmd)
    subprocess.CalledProcessError: Command '['cmake', '--build', '.', '--config', 'Release', '--', '_core.so']' returned non-zero exit status 2.
```

</p>
</details>

### CODE REVIEW

**Comment by okapies** [comment]
Jenkins, test this please.

File: tests/chainerx_tests/conftest.py [hunk 1]
```diff
@@ -123,14 +123,17 @@ def is_module(request):
     (2, 0, 3),
 ]
 
-_shapes_as_tuple_or_int = _shapes + [0, 1, 5]
+_shapes_as_sequence_or_int = (
+    _shapes
+    + list(map(lambda t: list(t), _shapes))
```
Comment (niboshi) [inline tests/chainerx_tests/conftest.py]: I don't think we need `list` conterparts for all the shapes.
How about adding a single or two parameters?

File: tests/chainerx_tests/unit_tests/routines_tests/test_manipulation.py [hunk 2]
```diff
@@ -394,6 +394,9 @@ def forward_xp(self, inputs, xp):
     ((3,), (2,)),
     ((3,), (3, 2)),
     ((1, 3), (3, 2)),
+    ((3,), [2]),
+    ((3,), [3, 2]),
+    ((1, 3), [3, 2]),
```
Comment (niboshi) [inline tests/chainerx_tests/unit_tests/routines_tests/test_manipulation.py]: Ditto.
I think adding a single configuration should be enough.

File: chainer/initializers/__init__.py [hunk 3]
```diff
@@ -37,7 +37,7 @@ def generate_array(initializer, shape, xp, dtype=None, device=None):
     Args:
         initializer: A callable object that takes :ref:`ndarray` and edits its
             value.
-        shape (tuple): Shape of a return array.
+        shape (int or tuple of int): Shape of a return array.
```
Comment (niboshi) [inline chainer/initializers/__init__.py]: Please separate this change, as it could be backported.

Comment (okapies) [reply chainer/initializers/__init__.py]: Moved to #7450.

**Comment by niboshi** [comment]
Jenkins, test this please

**Comment by niboshi** [comment]
Jenkins, test this please

**Comment by niboshi** [comment]
LGTM

**Pull request merged**